src/registry/http.ts

```diff
@@ -0,0 +1,453 @@
+      },
+    });
+  }
+
+  async authenticateBearer(ctx: AuthContext): Promise<HTTPContext> {
+    const params = new URLSearchParams({
+      service: ctx.service,
+      // explicitely include that we don't want an offline_token.
+      scope: `repository:${this.url.pathname.slice(1)}/image:pull,push`,
+      client_id: "r2registry",
+      grant_type: "password",
+      password: this.password(),
+    });
+    let res = await fetch(ctx.realm, {
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded",
+        "User-Agent": "Docker-Client/24.0.5 (linux)",
+      },
+      method: "POST",
+      body: params.toString(),
+    });
+    if (res.status === 404 || res.status === 405) {
+      console.debug(
+        this.url.toString(),
+        "Oauth 404/405... Falling back to simple token authentication, see https://distribution.github.io/distribution/spec/auth/token",
+      );
+      res = await this.authenticateBearerSimple(ctx, params);
+    }
+
+    if (!res.ok) {
+      throw new Error(
+        `unexpected ${res.status} from ${this.url.toString()} when Oauth authenticating: ${await res.text()}`,
+      );
+    }
+
+    const t = await res.text();
+    try {
+      const response: {
+        access_token?: string;
+        expires_in: number;
+        repository: string;
+        token?: string;
+      } = JSON.parse(t);
+
+      console.debug(
+        `Authenticated with registry ${this.url.toString()} successfully, got token that expires in ${
+          response.expires_in
+        } seconds`,
+      );
+
+      if (!response.access_token && !response.token) {
+        console.error(
+          "Oauth response doesn't have access_token field, doing fallback to password_env, however this might mean that we will 401 later",
+        );
+      }
+
+      return {
+        authContext: ctx,
+        repository: response.repository ?? this.url.pathname,
+        accessToken: response.access_token ?? response.token ?? this.authBase64(),
+      };
+    } catch (err) {
+      console.error(
+        "Doing json response in authentication: ",
+        errorString(err),
+        t.slice(0, Math.min(t.length, 100)),
+        "status",
+        res.status,
+      );
+      throw err;
+    }
+  }
+
+  async authenticateBasic(ctx: AuthContext): Promise<HTTPContext> {
+    const res = await fetch(ctx.realm, {
+      headers: {
+        Authorization: "Basic " + this.authBase64(),
+      },
+    });
+
+    if (!res.ok) {
+      throw new Error(`couldn't authenticate with registry ${ctx.realm}: ${JSON.stringify(await res.json())}`);
+    }
+
+    return {
+      authContext: ctx,
+      accessToken: this.authBase64(),
+      repository: this.url.pathname.slice(1),
+    };
+  }
+
+  async manifestExists(namespace: string, tag: string): Promise<CheckManifestResponse | RegistryError> {
+    try {
+      const ctx = await this.authenticate();
+      const req = ctxIntoRequest(ctx, this.url, "HEAD", `${namespace}/manifests/${tag}`);
```

Comment:
Maybe `req.headers.append("Accept", manifestTypes.join(", "));` is also needed here